base/setup/usetup/usetup.c

```diff
@@ -626,6 +626,173 @@ UpdateKBLayout(VOID)
+                                            NULL,
+                                            MUIGetString(STRING_REBOOTPROGRESSBAR),
+                                            ProgressTimeOutStringHandler);
+
+    StartTime = NtGetTickCount();
+    CONSOLE_Flush();
+
+    OldTimerValue = TimerValue;
+    while (TRUE)
+    {
+        /* Decrease the timer */
+        --TimeOut;
+
+        /*
+         * Compute how much time the previous operations took.
+         * This allows us in particular to take account for any time
+         * elapsed if something slowed down, or if the console has been
+         * paused in the meantime.
+         */
+        TimeElapsed = NtGetTickCount() - StartTime;
+        if (TimeElapsed >= 1000)
+        {
+            /* Increase StartTime by steps of 1 second */
+            TimeElapsed /= 1000;
+            StartTime += (1000 * TimeElapsed);
+
+            if (TimeElapsed <= TimerValue)
+                TimerValue -= TimeElapsed;
+            else
+                TimerValue = 0;
+
+            RefreshProgress = TRUE;
+        }
+
+        if (RefreshProgress)
+        {
+            ProgressSetStep(RebootProgressBar, OldTimerValue - TimerValue);
+            RefreshProgress = FALSE;
+        }
+
+        /* Stop when the timer reaches zero */
+        if (TimerValue <= 0)
+            break;
+
+        /*
+         * If the timer is used, use a passive wait of maximum 1 second
+         * while monitoring for incoming console input events, so that
+         * we are still able to display the timing count.
+         * Indeed, ReadConsoleInputW() indefinitely waits until an input
+         * event appears. ReadConsoleInputW() is however used to retrieve
+         * the input events where there are some, as well as for waiting
+         * indefinitely in case we do not use the timer.
+         */
+
+         /* Wait a maximum of 1 second for input events */
+        TimeElapsed = NtGetTickCount() - StartTime;
+        if (TimeElapsed < 1000)
+        {
+            /* Convert the time to NT Format */
+            Timeout.QuadPart = (1000 - TimeElapsed) * -10000LL;
+            Status = NtWaitForSingleObject(StdInput, FALSE, &Timeout);
+        }
+        else
+        {
+            Status = STATUS_TIMEOUT;
+        }
+
+        /* Check whether the input event has been signaled, or a timeout happened */
+        if (Status == STATUS_TIMEOUT)
+            continue;
+        if (Status != STATUS_WAIT_0)
+        {
+            /* An error happened, bail out */
+            DPRINT1("NtWaitForSingleObject() failed, Status 0x%08lx\n", Status);
+            break;
+        }
+
+        /* Check for an ENTER key press */
+        while (CONSOLE_ConInKeyPeek(Ir))
+        {
+            if (Ir->Event.KeyEvent.uChar.AsciiChar == 0x0D) /* ENTER */
+            {
+                /* Found it, stop waiting */
+                goto Stop;
+            }
+        }
+    }
+
+Stop:
```

Comment:
This is usually called `Exit`

Reply:
I think the "Stop" was used as "Stop countdown".